### PR TITLE
AuthZ: add headers for IP range AC checks for cloud data sources

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Add headers to outgoing DS requests that will be used for IP range verification.

**Why do we need this feature?**

To implement IP range access control for Grafana's cloud data sources.

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/identity-access-team/issues/363

**Special notes for your reviewer:**

This PR is not complete, as we're still debating how to sign the headers. Raising it early to share the current work with others.

### CODE REVIEW

**Review by aarongodin — COMMENTED** [review]
Looks good, just a couple minor things.

File: conf/defaults.ini [hunk 1]
```diff
@@ -429,6 +429,14 @@ default_home_dashboard_path =
...
 datasource_limit = 5000
 
+[datasources.ip_range_security]
+# Set to true to enable IP range access control for Grafana Cloud data sources
+enabled = false
+# Populated by Grafana Cloud. Do not edit.
+allow_list =
+# Populated by Grafana Cloud. Do not edit.
+secret_key =
+
```
Comment (Jguer) [inline conf/defaults.ini:439]: I don't think you need to add it to defaults.ini if it's not user facing

Comment (IevaVasiljeva) [reply conf/defaults.ini:439]: Thanks for the correction. I wasn't sure if we should add all settings to the ini file or not. Will remove it.

File: pkg/services/ngalert/accesscontrol.go [hunk 2]
```diff
@@ -25,6 +25,10 @@ var (
 					Action: accesscontrol.ActionAlertingRuleExternalRead,
 					Scope:  datasources.ScopeAll,
 				},
+				{
+					Action: "new:action",
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/ngalert/accesscontrol.go:29]: Whoops, yes, this was me testing something else 🙈

File: pkg/services/pluginsintegration/clientmiddleware/grafana_request_id_header_middleware.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	// if request is not for a datasource, skip the middleware
+	if h == nil || pCtx.DataSourceInstanceSettings == nil {
+		return
+	}
+
+	// Check if the request is for a datasource that is allowed to have the header
+	url := pCtx.DataSourceInstanceSettings.URL
+	if !slices.Contains(m.cfg.IPRangeACAllowedURLs, url) {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/pluginsintegration/clientmiddleware/grafana_request_id_header_middleware.go]: Good point, I've added `path.Clean` to the comparison. Do you think this is enough?

File: pkg/setting/setting.go [hunk 4]
```diff
@@ -1933,6 +1939,19 @@ func (cfg *Cfg) readDataSourcesSettings() {
 	cfg.DataSourceLimit = datasources.Key("datasource_limit").MustInt(5000)
 }
 
+func (cfg *Cfg) readDataSourceSecuritySettings() {
+	datasources := cfg.Raw.Section("datasources.ip_range_security")
+	cfg.IPRangeACEnabled = datasources.Key("enabled").MustBool(false)
+	cfg.IPRangeACSecretKey = datasources.Key("secret_key").MustString("")
+	allowedURLString := datasources.Key("allow_list").MustString("")
+	for _, url := range strings.Split(allowedURLString, ",") {
```
Comment (marefr) [inline pkg/setting/setting.go:1947]: Nit. I think we have a string split function in util package that you might want to use. Think we use that for parsing other settings that's delimited.

**Review by marefr — COMMENTED** [review]
FYI this will forward headers to backend plugins, I.e. api/ds/query. For requests using datasource proxy api/datasources/uid/:uid/proxy this will not apply - is that intended/expected?

File: pkg/services/pluginsintegration/clientmiddleware/grafana_request_id_header_middleware.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aarongodin) [reply pkg/services/pluginsintegration/clientmiddleware/grafana_request_id_header_middleware.go]: Nice, that makes sense to me 👍

**Comment by IevaVasiljeva** [comment]
> FYI this will forward headers to backend plugins, I.e. api/ds/query. For requests using datasource proxy api/datasources/uid/:uid/proxy this will not apply - is that intended/expected?

Is there any documentation on which endpoints are used when? I want to pass the headers on to any requests that target Grafana cloud data sources (eg, Loki, Mimir etc). So I guess if some requests to those pass through `api/datasources/uid/:uid/proxy` I will also need to change that?

**Comment by marefr** [comment]
> Is there any documentation on which endpoints are used when? I want to pass the headers on to any requests that target Grafana cloud data sources (eg, Loki, Mimir etc). So I guess if some requests to those pass through `api/datasources/uid/:uid/proxy` I will also need to change that?

Yes, that's correct. For outgoing HTTP requests (datasource proxy and backend plugins creating an HTTP client in core Grafana via pkg/infra/httpclient/provider.go) you might want to implement an HTTP client middleware https://github.com/grafana/grafana/tree/main/pkg/infra/httpclient/httpclientprovider that forwards your headers in outgoing HTTP requests.

**Review by Jguer — APPROVED** [review]

**Review by aarongodin — APPROVED** [review]

**Comment by IevaVasiljeva** [comment]
Thanks for the reviews all! I'll merge the PR now, so that we can start testing the e2e flow. Will cover the other endpoints in separate PRs.

**Pull request merged**